src/Microsoft.OData.Core/ODataBatchUtils.cs

```diff
@@ -142,5 +146,121 @@ internal static void EnsureArraySize(ref byte[] buffer, int numberOfBytesInBuffe
+        /// <param name="dependsOnRequestIds">Enumeration of request Ids used to lookup the request Id reference.</param>
+        /// <param name="baseUri">The baseUri used for validation.</param>
+        private static void ValidateReferenceUriForJsonBatch(Uri uri, IEnumerable<string> dependsOnRequestIds, Uri baseUri)
+        {
+            Debug.Assert(uri != null, "uri != null");
+
+            if (UriUtils.UriToString(uri).IndexOf('$') == -1)
+            {
+                // uri does not use $requestId,
+                // or request is in MultipartMixed format where  dependsOn is not required for reference Uri validation.
+                return;
+            }
+
+            string relativePath = null;
+            if (uri.IsAbsoluteUri)
+            {
+                if (baseUri == null)
+                {
+                    // The absolution Uri can contain $ character followed by name of system resource
+                    // such as $all, $metadata, etc, along with an unspecified baseUri. In such cases there are
+                    // not reference Uri to validate.
+                    return;
+                }
+
+                string baseUriString = UriUtils.UriToString(baseUri);
+                if (!uri.AbsoluteUri.StartsWith(baseUriString, StringComparison.Ordinal))
+                {
+                    throw new ODataException(Strings.ODataBatchReader_AbsoluteURINotMatchingBaseUri(
+                        uri.AbsoluteUri, baseUriString));
+                }
+
+                relativePath = uri.AbsoluteUri.Substring(baseUriString.Length);
+            }
+            else
+            {
+                relativePath = UriUtils.UriToString(uri);
+            }
+
+            // Trim the starting forward slashes as needed.
+            while (relativePath.StartsWith("/", StringComparison.Ordinal))
+            {
+                relativePath = relativePath.Substring(1);
+            }
+
+            if (relativePath.Length > 0 && relativePath[0] == '$')
+            {
+                // If the relative path starts with '$', find the request id reference optionally ended by '/'.
+                int idx = relativePath.IndexOf('/', 1);
+
+                string referenceId = idx > 0
+                    ? relativePath.Substring(1, idx - 1)
+                    : relativePath.Substring(1);
+
+                // Validate that dependsOn contains the Id reference.
+                // Note that the dependsOnRequestIds here is already flattened and consists of individual request
+                // ids, therefore it should contain the request id referenced by the Uri.
+                if (dependsOnRequestIds == null || !dependsOnRequestIds.Contains(referenceId))
+                {
+                    throw new ODataException(Strings.ODataBatchReader_ReferenceIdNotIncludedInDependsOn(
+                        referenceId, UriUtils.UriToString(uri), dependsOnRequestIds));
+                }
+            }
+        }

```

Comment:
ODataBatchUtils should be format-independent.
Validating the reference uri is a format-independent task, it's just that the set of valid ids are different between json and multipart/mixed.
In Json, the valid ids have to be called out as dependencies. In multipart/mixed, they are previous content-ids (in v4, they must be in the same batch; in v4.01 they can be any other content ids in the batch).

ODataBatchUtils should take the Uri, the set of valid references, and BaseUri.  The JsonWriter should call with the specified dependency ids; the multipart/mixed should call with the contentids of previous statements in the changeset (v4) or batch (v4.01). ODataBatchUtils should have no references to JSON or MultipartMixed derived classes.

Reply:
Updates have been made related to using BaseUri directly and passing in batchFormat, so that the coupling to the format-specific namespaces are reduced.
However I think the batchFormat is still needed. This method is invoked from a ODataBatchReader/Writer which are base classes for both formats,  therefore we cannot tell either formats apart. On the other hand we need to consider the backward compatibility to the original mulipart/mixed format where the dependsOnIds are not specified and shouldn't be rejected.

---
In reply to: [155347419](https://github.com/OData/odata.net/pull/992#discussion_r155347419) [](ancestors = 155347419)

Reply:
How about this:
1) Remove batchFormat from ODataBatchUtils.ValidateReferenceUri.
2) Replace the third parameter in ODataBatchUtils.ValidateReferenceUri with MessageReaderSettings (we may need additional information from MessageReaderSettings besides baseUri to properly validate in the future).
3) In BuildOperationRequestMessage, replace the "batchFormat" parameter with a Function that takes a requestUri, IEnumerable<string> dependency ids, and MessageReaderSettings (for extensibility this should probably actually be a dependency-injected ValidateUriReferenceService, but for now we can just make it a Func).
4) When called from Json, pass in ODataBatchUtils.ValidateReferenceUri.
5) When called from Multipart/Mixed, pass in an in-line defined Func that checks to see if dependencyIds are null and, if not, calls ODataBatchUtils.ValidateReferenceUri.
6) Same for BuildOperationRequestMessage in ODataBatchWriter

That way:
1) format-specific logic continues to be in the derived classes
2) multipart-mixed can retain existing non-validating behavior
3) we have extensibility for plugging in other validation logic

A cheaper version, if we don't think we'll want the ability to plug in custom validation logic, is to simply pass in a boolean ShouldValidateReferencedIds to BuildOperationRequestMessage.


---
In reply to: [156200555](https://github.com/OData/odata.net/pull/992#discussion_r156200555) [](ancestors = 156200555,155347419)

Reply:
As discussed offline with Mike, dependsOnIds for multipartMixed will include changes to enable tracking top-level statement for both V4 and V401 (and beyond). And this will be worked on after the 7.4.0-beta2 release. 

---
In reply to: [156797673](https://github.com/OData/odata.net/pull/992#discussion_r156797673) [](ancestors = 156797673,156200555,155347419)

Reply:
Do we have an issue to track this work if we close the comment?

---
In reply to: [157276865](https://github.com/OData/odata.net/pull/992#discussion_r157276865) [](ancestors = 157276865,156797673,156200555,155347419)

Reply:
I just created issue #1102  to track this.

---
In reply to: [157300821](https://github.com/OData/odata.net/pull/992#discussion_r157300821) [](ancestors = 157300821,157276865,156797673,156200555,155347419)

Reply:
Excellent; thanks!

---
In reply to: [157302502](https://github.com/OData/odata.net/pull/992#discussion_r157302502) [](ancestors = 157302502,157300821,157276865,156797673,156200555,155347419)